src/gpu/generic/sycl/rnn/ref_rnn.cpp

```diff
@@ -0,0 +1,791 @@
+/*******************************************************************************
+* Copyright 2019-2024 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*******************************************************************************/
+
+// General architecture
+//
+// for diff states, we have n_states + 1 as we have n_states diff
+// to propagate to the previous iteration and 1 states to propagate
+// to the previous layer
+// index 0 is dh for cell(t-1, l) to consume
+// index 1 is dc for cell(t-1, l) to consume
+// index 2 is dh for cell(t, l-1) to consume
+// this indexing enables to have the same indexing for states in elemwise
+// function
+// only the cell execution function should be impacted
+
+#include "gpu/generic/sycl/rnn/ref_rnn.hpp"
+#include "common/primitive.hpp"
+#include "common/primitive_desc.hpp"
+
+#include "common/matmul_pd.hpp"
+#include "common/stream.hpp"
+#include "common/type_helpers.hpp"
+#include "gpu/generic/sycl/rnn/rnn_kernels.hpp"
+
+#include <memory>
+
+#define DPRINT(fmt, ...) \
+    do { \
+        if (get_verbose_dev_mode(verbose_t::debuginfo) >= 2) { \
+            printf(fmt, __VA_ARGS__); \
+            fflush(nullptr); \
+        } \
+    } while (0)
+
+namespace dnnl {
+namespace impl {
+namespace gpu {
+namespace generic {
+namespace sycl {
+
+using namespace dnnl::impl::utils;
+using namespace dnnl::impl::math;
+using namespace prop_kind;
+using namespace alg_kind;
+using namespace rnn_utils;
+using namespace dnnl::impl::memory_tracking::names;
+
+#define AOC array_offset_calculator
+
+static status_t init_sycl(const rnn_pd_t *rnn_pd, const rnn_utils::conf_t &rnn,
+        const memory_desc_wrapper &src_layer_d,
+        const memory_desc_wrapper &src_iter_d,
+        const memory_desc_wrapper &weights_layer_d,
+        const memory_desc_wrapper &weights_iter_d,
+        const memory_desc_wrapper &bias_d,
+        const memory_desc_wrapper &dst_layer_d,
+        const memory_desc_wrapper &dst_iter_d, const memory_desc_wrapper &ws_d,
```

Comment:
We can avoid passing all the  `memory_desc_wrapper_t` objects as they can be extracted from `rnn_pd`, similar to what is done in 1e7d50e68d6c5f86778f2d11cabb85e97d48a3d2.

Reply:
Removed `memory_desc_wrapper_t`s and now passing the `rnn_pd` isntead.